### PR TITLE
Enumerable methods (Ruby practice)

### PR DESCRIPTION
**Pull request by frederico-miranda**
This is a pull request opened for receiving a code review. It's a practice for the Ruby programming language which consist of implementing again the methods of the Enumerable module. The path for the file that contains the code for this practice is _/my-enumerable/my_enumerable.rb_

The general requirements for this project to be considered valid can be read at the following link:
[The Odin Project - Enumerable Methods](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#project-2-enumerable-methods)

### CODE REVIEW

File: my-enumerable/my_enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+  def my_any?(pattern = nil)
+    passes = false
+    my_each do |value|
+      passes ||= my_aux(pattern, value) { |x| block_given? ? (yield x) : x }
+    end
+    passes
+  end
+
+  def my_none?(pattern = nil)
```
[the remark this answers is not in the document]
Comment (frederico-miranda) [reply my-enumerable/my_enumerable.rb]: This has been addressed. Now #my_count accepts a block.

File: my-enumerable/my_enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    !my_any?(pattern) { |x| block_given? ? (yield x) : x }
+  end
+
+  def my_count
+    return length if is_a?(Array)
+
+    return keys.length if is_a?(Hash)
+  end
+
+  def my_map(given_proc = nil)
```
[the remark this answers is not in the document]
Comment (frederico-miranda) [reply my-enumerable/my_enumerable.rb]: This has been addressed. The misnamed variable has been properly renamed.

File: my-enumerable/my_enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    base = []
+    if !_proc.nil?
+      my_each_with_index { |value| base.push(given_proc.call(value)) }
+    else
+      my_each_with_index { |value| base.push(yield value) }
+    end
+    base
+  end
+
+  def my_inject(initial = nil)
```
[the remark this answers is not in the document]
Comment (frederico-miranda) [reply my-enumerable/my_enumerable.rb]: This has been addressed. The method #my_inject now has a guard case for when there is no block given.

File: my-enumerable/my_enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    false
+  end
+
+  def my_all?(pattern = nil)
+    passes = true
+    my_each do |value|
+      passes &&= my_aux(pattern, value) { |x| block_given? ? (yield x) : x }
+    end
+    passes
+  end
```
Comment (imhta) [inline my-enumerable/my_enumerable.rb]: ### Your `my_all` method failing some test
- [ ]  enumerables my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] enumerables my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: my-enumerable/my_enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+  def my_any?(pattern = nil)
+    passes = false
+    my_each do |value|
+      passes ||= my_aux(pattern, value) { |x| block_given? ? (yield x) : x }
+    end
+    passes
+  end
```
Comment (imhta) [inline my-enumerable/my_enumerable.rb]: ### Your `my_any` method failing some test
- [ ] enumerables my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class

File: my-enumerable/my_enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+  def my_none?(pattern = nil)
+    !my_any?(pattern) { |x| block_given? ? (yield x) : x }
+  end
```
Comment (imhta) [inline my-enumerable/my_enumerable.rb]: ### Your `my_none` method failing some test
- [ ] enumerables my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ]  enumerables my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: my-enumerable/my_enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+  def my_map(given_proc = nil)
+    base = []
+    if !given_proc.nil?
+      my_each_with_index { |value| base.push(given_proc.call(value)) }
+    else
+      my_each_with_index { |value| base.push(yield value) }
+    end
+    base
+  end
```
Comment (imhta) [inline my-enumerable/my_enumerable.rb]: ### Your `my_map` method failing some test

- [ ] enumerables my_map returns an Enumerator if no block is given

File: my-enumerable/my_enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    if initial.nil?
+      flat_self = to_a
+      initial = flat_self.shift
+      flat_self.my_each { |value| initial = yield initial, value }
+    else
+      my_each { |value| initial = yield initial, value }
+    end
+    initial
+  end
+end
```
Comment (imhta) [inline my-enumerable/my_enumerable.rb]: ### Your `my_inject` method failing some test
- [ ] enumerables my_inject combines all elements of enum by applying a binary operation, specified by a block
- [ ] enumerables my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

Comment (frederico-miranda) [reply my-enumerable/my_enumerable.rb]: > by applying a binary operation, specified by a block
I believe I have addressed this issue. The ```my_inject```method now support blocks, as corroborated by the following short test:

```Ruby
[3,4,5,6,7,8].inject { |a,b| a + b }
=> 33

[3,4,5,6,7,8].my_inject { |a,b| a + b }
=> 33
```

> by applying the symbol as a named method
I believe I have addressed this issue. The ```my_inject``` method now supports symbols as the input, as the following short test corroborates this:

```Ruby
[3,4,5,6,7,8].inject(:*)
=> 20160

[3,4,5,6,7,8].my_inject(:*)
=> 20160
```

File: my-enumerable/my_enumerable.rb (hunk 7 above)
Comment (frederico-miranda) [reply my-enumerable/my_enumerable.rb]: > enumerables my_map returns an Enumerator if no block is given
That has been addressed. I made a quick test for this:

```Ruby
[3,4,5,6,7,8].map
=> #<Enumerator: [3, 4, 5, 6, 7, 8]:map>
[3,4,5,6,7,8].my_map
=> #<Enumerator: [3, 4, 5, 6, 7, 8]:each>
```

File: my-enumerable/my_enumerable.rb (hunk 6 above)
Comment (frederico-miranda) [reply my-enumerable/my_enumerable.rb]: I have addressed these issues by fixing the  ```my_any?``` method.

```Ruby
[3,4,5,2, "abc"].none?(nil.class)
=> true
[3,4,5,2, "abc"].my_none?(nil.class)
=> true
[3,4,5,2, "abc"].none?(String)
=> false
[3,4,5,2, "abc"].my_none?(String)
=> false
[3, 4, 5, 2].none? "abc"
=> true
[3, 4, 5, 2].my_none? "abc"
=> true
[3, 4, 5, 2].none? 2
=> false
[3, 4, 5, 2].my_none? 2
=> false
[3, 4, 5, 2].none?
=> false
[3, 4, 5, 2].my_none?
=> false
```

File: my-enumerable/my_enumerable.rb (hunk 5 above)
Comment (frederico-miranda) [reply my-enumerable/my_enumerable.rb]: > when a class is passed as an argument returns true if at least one of the collection is a member of such class
I have addressed that. Now it's possible to use a class for checking if any members of the  collection is of that class. I made a quick test:

```
[3,4,5,2, "abc"].any?(nil.class)
=> false
[3,4,5,2, "abc"].my_any?(nil.class)
=> false
[3,4,5,2, "abc"].any?(String)
=> true
[3,4,5,2, "abc"].my_any?(String)
=> true
```

File: my-enumerable/my_enumerable.rb (hunk 4 above)
Comment (frederico-miranda) [reply my-enumerable/my_enumerable.rb]: That has been implemented. Now classes or reference values can be used in ```my_all?``` method. I have made a quick test for this:

```Ruby
[2, 2, 2, 2, 2, 2, "abc"].all? 2
=> false
[2, 2, 2, 2, 2, 2, "abc"].my_all? 2
=> false
[2, 2, 2, 2, 2, 2, "abc"].all? Integer
=> false
[2, 2, 2, 2, 2, 2, "abc"].my_all? Integer
=> false
 [2, 2, 2, 2, 2, 2].all? 2
=> true
[2, 2, 2, 2, 2, 2].my_all? 2
=> true
[2, 2, 2, 2, 2, 2, "abc"].all? 2
=> false
[2, 2, 2, 2, 2, 2, "abc"].my_all? 2
=> false

```

**Comment by Anna-Myzukina** [comment]
#### Pull Request under the review by Anna Muzykina (TSE)

**Comment by Anna-Myzukina** [comment]
### Hi Frederico!  
#### Wow, as I can see now all methods working correctly:+1: 

#### Project approved!

*Happy coding 😄*

**Pull request merged**